lib/evmone/tracing.hpp

```diff
@@ -0,0 +1,49 @@
+// evmone: Fast Ethereum Virtual Machine implementation
+// Copyright 2021 The evmone Authors.
+// SPDX-License-Identifier: Apache-2.0
+#pragma once
+
+#include <evmc/instructions.h>
+#include <memory>
+#include <string_view>
+
+namespace evmone
+{
+using bytes_view = std::basic_string_view<uint8_t>;
+
+class Tracer
+{
+    friend class VM;  // Has access the the m_next_tracer to traverse the list forward.
+    std::unique_ptr<Tracer> m_next_tracer;
+
+public:
+    virtual ~Tracer() = default;
+
+    void notify_execution_start(bytes_view code) noexcept  // NOLINT(misc-no-recursion)
+    {
+        on_execution_start(code);
+        if (m_next_tracer)
+            m_next_tracer->notify_execution_start(code);
+    }
+
+    void notify_execution_end() noexcept  // NOLINT(misc-no-recursion)
+    {
+        on_execution_end();
+        if (m_next_tracer)
+            m_next_tracer->notify_execution_end();
+    }
+
+    void notify_instruction_start(uint32_t pc) noexcept  // NOLINT(misc-no-recursion)
```

Comment:
Ah, so it seems that with `notify_instruction_start` and `on_execution_end`, there wouldn't be a need for `notify_instruction_end`.  I'm curious if there are any awkward edge-cases for an EIP3155-compatible tracer because of this.

Reply:
We can also add `notify_instruction_end()` if needed. The biggest issue seems to be "instruction gas cost".